### PR TITLE
Configure widget

### PR DESCRIPTION
**Pull request by kaareal**
The docs/stream.html is still somewhat a draft and i am thinking its full of spelling errors. My plan is to sit down tomorrow with justin or johan. And complete it. 

But in the meantime check the rest out.

Basically the major difference is now that the mapping is attached directly on the stream object.

@Juice10 @drakonen

### CODE REVIEW

File: serve/unit/api.js [hunk 1]
```diff
@@ -20,12 +20,13 @@ define([
         };
         var callback = function(){};
 
-        API.getUrl(url, params, callback, {disableCache: true});
+        API.getUrl(url, params, callback, {disableCache: true, apiHost: 'http://v3.streams.bottlenose.com/3'});
```
Comment (Juice10) [inline serve/unit/api.js]: why the <code>**v3.**streams.bottlenose.com/3</code>?

File: serve/unit/mapping-processor.js [hunk 2]
```diff
@@ -1,61 +1,82 @@
...
+  'backbone',
+], function(Openhose, MappingProcessor, Backbone) {
+
   describe('MappingProcessor', function(){
-    describe('metricsById', function(){
-      this.afterEach(function () {
-        Openhose.configureDefaults();
+    var server;
+
+    beforeEach(function () {
```
Comment (Juice10) [inline serve/unit/mapping-processor.js]: I believe you have to do `this.beforeEach` and `this.afterEach` otherwise it leaks into all tests.  (not sure why, but it just does, have had a lot of headaches with it)

File: docs/stream.html [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+        organizationToken: "fb74ed08c3669...",
+        organizationId: "54c22ac91342df12d0f2a288",
+        mappingUrl: "http://streams.bottlenose.com/3/mappings/social-media"
+      }</b>,
+      ...
+    });
+  </pre>
+
+  <p>
+    In the above examples the stream options are passed directly into the widget. And not shared with other widgets. Like the global <b>Openhose.configure</b> stream options would be.
```
Comment (drakonen) [inline docs/stream.html]: Like the global <b>Openhose.configure</b> stream options would be
Unlike the global <b>Openhose.configure</b> stream options would be

File: docs/stream.html [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+      ...
+    });
+  </pre>
+
+  <p>
+    In the above examples the stream options are passed directly into the widget. And not shared with other widgets. Like the global <b>Openhose.configure</b> stream options would be. 
+  </p>
+  <p>
+    The <b>stream id</b> and <b>organizationId</b> and <b>organizationToken</b> can be found inside 
+    <a href="//bottlenose.com/tools/api" target="_blank">bottlenose.com/tools/api</a> (note that naming is slight different organizationToken equal userToken, organizationId equal userId)
```
Comment (drakonen) [inline docs/stream.html]: slight different organizationToken
slightly different: organizationToken

File: docs/stream.html [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+  </p>
+
+  <p>
+    The <b>mappingUrl</b> is used to fetch the mapping which is used to describe the data contained in the stream. How the mapping looks for bottlenose.com can be found here (<a href="//bottlenose.com/proxy/openhose/3/mappings" target="_blank">bottlenose.com/proxy/openhose/3/mappings</a>). For a social media stream the mapping looks like this <a href="//streams.bottlenose.com/3/mappings/social-media" target="_blank">streams.bottlenose.com/3/mappings/social-media</a>.
+  </p>
+  <p> 
+    The <b>mapping</b> of a stream can also be specify manually, instead of using the <b>mapingUrl</b>. By setting a mapping attribute on the stream and pass into the widget.
```
Comment (drakonen) [inline docs/stream.html]: What happens when you use both `mappingUrl` and `mapping`?

File: docs/stream.html [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+                    volume: {
+                      label: 'My Volume Label'
+                    },
+                    impressions: {
+                      label: 'My Impression Label'
+                    }
+                  }
+                }
+              },
+              viewOptions: {
```
Comment (drakonen) [inline docs/stream.html]: These viewoptions seem irrelevant to the example

File: serve/unit/mapping-processor.js [hunk 7]
```diff
@@ -1,61 +1,82 @@
...
+
+    beforeEach(function () { 
+      server = sinon.fakeServer.create();
+    });
+
+    afterEach(function(){
+      server.restore();
+    });
+
+    it('hasMapping', function() {
```
Comment (drakonen) [inline serve/unit/mapping-processor.js]: Wasn't the convention to use it('should ...')

**Comment by drakonen** [comment]
:+1: :jack_o_lantern:

File: serve/unit/api.js (hunk 1 above)
Comment (kaareal) [inline serve/unit/api.js]: because a bit later i check for it 
assert.equal(jQueryMock.args[0][0].url, 'http://v3.streams.bottlenose.com/3' + url);

i wanted a different domain than streams.bottlenose.com as the default might kick in.

File: serve/unit/mapping-processor.js (hunk 2 above)
Comment (kaareal) [inline serve/unit/mapping-processor.js]: good point

File: docs/stream.html (hunk 5 above)
Comment (kaareal) [inline docs/stream.html]: it uses the mapping then

File: src/lib/mapping-processor.js [hunk 8]
```diff
@@ -1,625 +1,54 @@
...
-  MappingProcessor.hasMapping = function (sourceType) {
-    return !!MAPPINGS[sourceType];
+  var MappingProcessor = {};
+  var cache = {
+    data: {}
   };
 
-  MappingProcessor.setMapping = function(mappings) {
-    MAPPINGS = mappings;
+  //checks if the stream already got a mapping available
```
Comment (Juice10) [inline src/lib/mapping-processor.js]: should be:
`// checks if the stream already has a mapping available`

File: src/lib/mapping-processor.js [hunk 9]
```diff
@@ -1,625 +1,54 @@
...
+  MappingProcessor.hasMapping = function (stream) {
+    var mapping = stream.get("mapping");
+    var mappingUrl = stream.get("mappingUrl");
+    if (!mapping && !mappingUrl) return true;
```
Comment (Juice10) [inline src/lib/mapping-processor.js]: shouldn't this return `false`?

File: src/processor/analytics-mixin.js [hunk 10]
```diff
@@ -321,20 +289,7 @@ define(['../core', '../lib/utils'], function(Openhose, Utils) {
...
-            info: metric,
-            value: this.metrics[0].total(metric.metricId)
-          };
-        });
-      } else {
-        segments = this._segmentsForMetrics(metrics, options);
-      }
-
-      return segments;
+      return this._segmentsForMetrics(metrics, options);
```
Comment (Juice10) [inline src/processor/analytics-mixin.js]: shouldn't we just get rid of `_tableSegmentsForMetrics` and rename all of the calls to `_segmentsForMetrics`

File: serve/unit/api.js (hunk 1 above)
Comment (Juice10) [inline serve/unit/api.js]: lets rename that to something more obvious then, like `http://mocked-domain.bottlenose.com/3`
